[MRG+2] warn_on_dtype for DataFrames

wdevazelhes opened the pull request:
#### Reference Issues/PRs
Fixes #10948 


#### What does this implement/fix? Explain your changes.

Pandas DataFrames can have different dtypes for each Series they contain. This PR fixes #10948  while displaying a message specific to this case of ``DataFrames`` with several dtypes. However this code just adds some "ifs" and the behaviour is maybe not what we would want. But if it is OK I will also add some tests. 

The following examples shows the new behaviour of this PR: 

#### Example 1 (the one in the issue):

```python
from sklearn.utils.validation import check_array
import pandas as pd
df = pd.DataFrame([[1, 2, 3], [2, 3, 4]], dtype=object)
checked = check_array(df, warn_on_dtype=True)
```

Returns: 

```python-traceback
DataConversionWarning: Data with input dtype object were all converted to float64.
  warnings.warn(msg, DataConversionWarning)
```

#### Example 2: some dtypes are different than the one asked

```python
from sklearn.utils.validation import check_array
import pandas as pd
df = pd.DataFrame([[1, 2, 3], [2, 3, 4]], dtype=object)
df[1] = df[1].astype(float)
df[2] = df[2].astype(int)
# df3 stays object
checked = check_array(df, warn_on_dtype=True) # asks for conversion to numerical type
```

Returns: 

```python-traceback
DataConversionWarning: Data with input dtype int64, float64, object were all converted to float64.
  warnings.warn(msg, DataConversionWarning)
```

#### Example 3: all dtypes are different than the one asked

```python
from sklearn.utils.validation import check_array
import pandas as pd
df = pd.DataFrame([[1, 2, 3], [2, 3, 4]], dtype=object)
df[1] = df[1].astype(float)
checked = check_array(df, dtype=['int'], warn_on_dtype=True) # asks for conversion to int
```

Returns: 

```python-traceback
DataConversionWarning: Data with input dtype float64, object were all converted to int64.
  warnings.warn(msg, DataConversionWarning)
```

Commit:
[WIP] fixes #10948

rth reviewed (commented):
Thanks for your PR! Overall this feature makes sense, I think. I haven't reviewed the code yet.

Please change the title to something meaningful (i.e. describe what this PR does). Also please add some unit tests that are skipped if pandas is not found (cf this [example](https://github.com/scikit-learn/scikit-learn/blob/afe540c7f2cbad259dd333e6744b088213180bee/sklearn/utils/tests/test_validation.py#L610-L619))

wdevazelhes commented:
Will do, thanks! Should I also make some tests with MockDataFrame ? (I should then add a ``dtypes`` attribute to MockDataFrame)

Commit:
fix travis test with pandas 0.20

jnothman commented:
you might be better off doing it with a real DataFrame, using importorskip
to avoid the test where pandas is unavailable

Commit:
Add test.

wdevazelhes commented:
Done, I also changed the title of the PR to MRG, since all tests pass

GaelVaroquaux commented on sklearn/utils/validation.py:
```diff
@@ -445,6 +445,12 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
+    # check if the object contains several dtypes (typically a pandas
+    # DataFrame), and store them. If not, store None.
+    dtypes_orig = None
+    if hasattr(array, "dtypes"):
```
Maybe we should be a bit more precise on the duck typing, to avoid to capture other objects that happen by chance to have a dtype attribute. For instance, we could test for "__array__" which, I believe is the method used by numpy to convert to a numpy array.

jnothman replied:
Note that the attribute here is dtypes, not dtype.

wdevazelhes replied:
Thanks, will do

GaelVaroquaux commented on sklearn/utils/validation.py:
```diff
@@ -556,6 +562,16 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
+        # some data must have been converted
+        msg = ("Data with input dtype %s were all converted to %s%s."
+               % (', '.join(map(str, set(dtypes_orig))), array.dtype, context))
+        warnings.warn(msg, DataConversionWarning)
```
Maybe add "staclevel=2" here, or should we have 3?

wdevazelhes replied:
Thanks, will do

GaelVaroquaux commented:
+1 on merge once the small comments have been addressed.

wdevazelhes commented:
Thanks for the review @GaelVaroquaux , I ll update the code according to your comments and resolve conflicts

Commit:
Merge branch 'master' into fix/check_array_df_conversion

# Conflicts:
#	sklearn/utils/validation.py

Commit:
ENH: add stacklevel=2 for better warning log

Commit:
ENH: add more precise duck typing (see comment https://github.com/scikit-learn/scikit-learn/pull/10949#discussion_r182351141)

Commit:
FIX: flake8 remove blank line

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -562,6 +568,15 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
+        # (for instance in a DataFrame that can contain several dtypes) then
+        # some data must have been converted
+        msg = ("Data with input dtype %s were all converted to %s%s."
+               % (', '.join(map(str, set(dtypes_orig))), array.dtype, context))
```
this does not have deterministic order.... which is probably fine. I would have preferred `'%s' % sorted(dtypes_orig)` but this will be alright.

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -562,6 +568,15 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
+        # some data must have been converted
+        msg = ("Data with input dtype %s were all converted to %s%s."
+               % (', '.join(map(str, set(dtypes_orig))), array.dtype, context))
+        warnings.warn(msg, DataConversionWarning, stacklevel=2)
```
The other DataConversionWarning does not use `stacklevel`. Generally, I find stacklevel unhelpful. (The best way to get the stack info correct is to run again with an `'error'` filter.)

jorisvandenbossche replied:
I personally find correct stacklevels very useful (typically when not at the interactive interpreter), however, since this is in `check_array` which is then still used in other sklearn methods (and not directly by the user), I agree this is less useful (or at least should be more than 2)

jnothman reviewed (approved):
But LGTM

jnothman commented:
Do we think this needs a what's new entry?

jnothman committed:
Merge branch 'master' into fix/check_array_df_conversion

jorisvandenbossche commented on sklearn/utils/validation.py:
```diff
@@ -581,6 +587,15 @@ def check_array(array, accept_sparse=False, accept_large_sparse=True,
         array = np.array(array, dtype=dtype, order=order)
 
+    if warn_on_dtype and dtypes_orig is not None and {array.dtype} != \
+            set(dtypes_orig):
```
@jnothman does sklearn have some style rules on avoiding `\` in new code for line continuation?

Commit:
FIX take into account review comments:
- sort dtypes to make output deterministic
- put parenthesis instead of backslash for continuation line
- put stacklevel=3 for more targeted output

Commit:
Merge branch 'fix/check_array_df_conversion' of https://github.com/wdevazelhes/scikit-learn into fix/check_array_df_conversion

wdevazelhes commented:
Thanks for the review @jnothman and @jorisvandenbossche 
I just updated the code with your comments

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -568,14 +568,15 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
 
-    if warn_on_dtype and dtypes_orig is not None and {array.dtype} != \
-            set(dtypes_orig):
+    if (warn_on_dtype and dtypes_orig is not None and {array.dtype} !=
```
I would much prefer the line break next to a Boolean operators than a comparison, given order of precedence!

wdevazelhes replied:
You're right
Done

Commit:
REF move line break after logical operator

jnothman commented:
Thanks!!​

jorisvandenbossche commented:
@wdevazelhes In https://github.com/scikit-learn/scikit-learn/pull/12104, @ogrisel is removing the warning in case the original data was object dtype data. 

In your original issue report, you used an object array as example:

```
df = pd.DataFrame([[1, 2, 3], [2, 3, 4]], dtype=object)
checked = check_array(df, warn_on_dtype=True)
```

did you have a specific use case for this (specifically for object dtype)? Or was this rather as a small example to reproduce the issue and was the actual case where you encountered this not about numeric object being casted to object?

jorisvandenbossche commented:
Another comment, now looking at the changes in this PR. I think a case where you have all numeric data in a DataFrame, but which is consisting of mixed int/float columns, is quite common. Now that is processed fine by sklearn transformers as this gets simply converted to a floating array. 

But with the change in this PR, that will start to raise a warning in eg the StandardScaler. Do we think it is actually worth to warn in such a case?

wdevazelhes commented:
> @wdevazelhes In #12104, @ogrisel is removing the warning in case the original data was object dtype data.
> 
> In your original issue report, you used an object array as example:
> 
> ```
> df = pd.DataFrame([[1, 2, 3], [2, 3, 4]], dtype=object)
> checked = check_array(df, warn_on_dtype=True)
> ```
> 
> did you have a specific use case for this (specifically for object dtype)? Or was this rather as a small example to reproduce the issue and was the actual case where you encountered this not about numeric object being casted to object?


I think this was just a small example to reproduce the issue, I could probably have put something else instead of object





> Another comment, now looking at the changes in this PR. I think a case where you have all numeric data in a DataFrame, but which is consisting of mixed int/float columns, is quite common. Now that is processed fine by sklearn transformers as this gets simply converted to a floating array.
> 
> But with the change in this PR, that will start to raise a warning in eg the StandardScaler. Do we think it is actually worth to warn in such a case?

In fact if I remember correctly, at that time I was trying to understand when `check_array` would return a copy of my data or not, and I used `warn_on_dtype` as a proxy for that (because when there is a conversion I think there was also a copy although I'm not sure it is always the case)
I agree that it might be too much to raise a warning if ints are converted into floats for instance, inside a transformer

qinhanmin2014 commented:
FYI seems that this introduces a bug #12622, suggestions are welcomed since we're going to include the fix in 0.20.1.

amueller commented:
fix in #12625
